crypto_box.js

```diff
@@ -95,6 +104,166 @@ function crypto_box_seal_open (m, c, pk, sk) {
   return crypto_secretbox_open_easy(m, c.subarray(epk.length), n, k)
 }
 
+// int
+// crypto_box_curve25519xsalsa20poly1305_beforenm(unsigned char *k,
+//                                                const unsigned char *pk,
+//                                                const unsigned char *sk)
+// {
+//     static const unsigned char zero[16] = { 0 };
+//     unsigned char s[32];
+//
+//     if (crypto_scalarmult_curve25519(s, sk, pk) != 0) {
+//         return -1;
+//     }
+//     return crypto_core_hsalsa20(k, zero, s, NULL);
+// }
+function crypto_box_beforenm (k, pk, sk) {
+  const zero = new Uint8Array(16)
+  const s = new Uint8Array(32)
+
+  if (crypto_scalarmult(s, sk, pk) !== 0) {
+    return -1
+  }
+
+  xsalsa20.core_hsalsa20(k, zero, s, xsalsa20.SIGMA)
+
+  return 0
```

Comment:
Ideally this should return `true` / `false` and then the assert the result: `assert(crypto_box_beforenm(k, pk, sk))`.

This deviates from libsodium for clearer error handling, since we assert inside the called function rather than checking return values